site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+
+## Widget lifecycle changes
+
+We were never completely happy with the APIs we provided for end users to interact with the widget lifecycle.  We have long felt that one of the biggest challenges with Dojo 1 Dijits was a complicated lifecycle.  In general the Dojo 2 widgeting system takes care of everything and developers really shouldn't need to worry about the lifecycle and instead just worry about rendering when requested.  There are still use cases where a developer does need to know when the virtual DOM they render is added to or removed from the DOM.
+
+In Beta 4 we have introduced two lifecycle methods to `WidgetBase`.  They are called `onAttach` and `onDetach`.  They are designed to allow a widget developer to _hook_ when a widget instance is attached to the DOM or detached from the DOM.
+
+To use them, it simply requires overriding the `onAttach` or `onDetach` methods.  For example:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+
+export default class MyWidget extends WidgetBase {
+    protected onAttach() {
+        // perform actions when attached
+    }
+
+    protected onDetach() {
+        // perform actions when detached
+    }
+
+    protected render() {
+        return v('div', { }, [ 'Hello world!' ])
+    }
+}
+```
+
+Previously, we had other capabilities which provided functionality in this area, but we have deprecated those and will remove them in the future.
+
+## Deferred properties
+
+A virtual DOM node can now provide its properties in a way that are resolved in a deferred fashion, where instead of providing an object of `VirtualDomProperties`, you can provide a function which takes a boolean flag of if the virtual DOM node has been inserted into the DOM and returns an object of `VirtualDomProperties`.
+
+Because these functions will be called during the next `requestAnimationFrame` (when the projector is running in its default asynchronous mode), this feature can be used to better handle CSS transition states.  For example, you can now do something like this:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+
+render() {
+    return v('div', (inserted) => {
+        return {
+            classes: 'fadeInTransition',
+            styles: {
+                opacity: inserted ? '1' : '0'
+            }
+        };
+    }, [ 'Hello World!' ]);
+}
+```
+
+## Web animations meta provider
+
+One of the goals of the Dojo 2 widgeting system is to provide sufficient tools and abstractions so that widget authors don't have to deal directly with the DOM.  Meta providers provide ways to interact with the DOM, sometimes in complex ways, without having to sweat the details.
```

Comment:
Meta providers provide ways to interact with the DOM, sometimes in complex ways, without having to sweat the details. -> Meta providers offer reactive ways to interact with features that would otherwise require direct DOM interaction. Dojo 2 includes meta providers for Intersection Observers, dragging on an element, node dimensions, and the matching of a virtual DOM node.